cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/GcpMetadataHttpHandlerInternal.java

```diff
@@ -123,29 +138,77 @@ public void token(HttpRequest request, HttpResponder responder,
               METADATA_FLAVOR_HEADER_KEY, METADATA_FLAVOR_HEADER_VALUE));
     }
 
-    // TODO: CDAP-20750
-    if (metadataServiceEndpoint == null) {
+    if (gcpMetadataTaskContext != null) {
+      try {
+        // fetch token from credential provider
+        GcpTokenResponse gcpTokenResponse =
+            Retries.callWithRetries(() -> fetchTokenFromCredentialProvider(scopes),
+                RetryStrategies.fromConfiguration(cConf, Constants.Service.TASK_WORKER + "."));
+        responder.sendJson(HttpResponseStatus.OK, GSON.toJson(gcpTokenResponse));
+        return;
+      } catch (NotFoundException e) {
+        // if credential identity not found,
+        // fallback to gcp metadata server for backward compatibility.
+      } catch (Exception ex) {
+        if (ex instanceof HttpErrorStatusProvider) {
+          HttpResponseStatus status = HttpResponseStatus.valueOf(
+              ((HttpErrorStatusProvider) ex).getStatusCode());
+          responder.sendJson(status, exceptionToJson(ex));
+        } else {
+          LOG.warn("Failed to fetch token from credential provider", ex);
+          responder.sendJson(HttpResponseStatus.INTERNAL_SERVER_ERROR, exceptionToJson(ex));
+        }
+        return;
+      }
+    }
+
+    if (metadataServiceTokenEndpoint == null) {
       responder.sendString(HttpResponseStatus.NOT_IMPLEMENTED,
           String.format("%s has not been set",
               Constants.TaskWorker.METADATA_SERVICE_END_POINT));
       return;
     }
 
     try {
-      URL url = new URL(metadataServiceEndpoint);
-      if (!Strings.isNullOrEmpty(scopes)) {
-        url = new URL(String.format("%s?scopes=%s", metadataServiceEndpoint, scopes));
-      }
-      io.cdap.common.http.HttpRequest tokenRequest = io.cdap.common.http.HttpRequest.get(url)
-          .addHeader(METADATA_FLAVOR_HEADER_KEY, METADATA_FLAVOR_HEADER_VALUE)
-          .build();
-      HttpResponse tokenResponse = HttpRequests.execute(tokenRequest);
-      responder.sendJson(HttpResponseStatus.OK, tokenResponse.getResponseBodyAsString());
+      responder.sendJson(HttpResponseStatus.OK,
+          fetchTokenFromMetadataServer(scopes).getResponseBodyAsString());
     } catch (Exception ex) {
       LOG.warn("Failed to fetch token from metadata service", ex);
-      responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, exceptionToJson(ex),
-          new DefaultHttpHeaders().set(HttpHeaders.CONTENT_TYPE, "application/json"));
+      responder.sendJson(HttpResponseStatus.INTERNAL_SERVER_ERROR, exceptionToJson(ex));
+    }
+  }
+
+  private GcpTokenResponse fetchTokenFromCredentialProvider(String scopes) throws NotFoundException,
```

Comment:
Can we move this to a remote implementation of the `CredentialProvider` class and inject that implementation in the sidecar?